src/wallet/asyncrpcoperation_common.cpp

```diff
@@ -40,3 +41,115 @@ std::pair<CTransaction, UniValue> SignSendRawTransaction(UniValue obj, std::opti
+        const InputSelectionError& err,
+        const ZTXOSelector& selector,
+        const TransactionStrategy& strategy)
+{
+    std::visit(match {
+        [](const AddressResolutionError& err) {
+            switch (err) {
+                case AddressResolutionError::SproutSpendNotPermitted:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "Sending from the Sprout shielded pool to the Sapling "
+                        "shielded pool is not enabled by default because it will "
+                        "publicly reveal the transaction amount. THIS MAY AFFECT YOUR PRIVACY. "
+                        "Resubmit with the `privacyPolicy` parameter set to `AllowRevealedAmounts` "
+                        "or weaker if you wish to allow this transaction to proceed anyway.");
+                case AddressResolutionError::SproutRecipientNotPermitted:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "Sending funds into the Sprout pool is no longer supported.");
+                case AddressResolutionError::TransparentRecipientNotPermitted:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "This transaction would have transparent recipients, which is not "
+                        "enabled by default because it will publicly reveal transaction "
+                        "recipients and amounts. THIS MAY AFFECT YOUR PRIVACY. Resubmit "
+                        "with the `privacyPolicy` parameter set to `AllowRevealedRecipients` "
+                        "or weaker if you wish to allow this transaction to proceed anyway.");
+                case AddressResolutionError::InsufficientSaplingFunds:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "Sending from the Sapling shielded pool to the Orchard "
+                        "shielded pool is not enabled by default because it will "
+                        "publicly reveal the transaction amount. THIS MAY AFFECT YOUR PRIVACY. "
+                        "Resubmit with the `privacyPolicy` parameter set to `AllowRevealedAmounts` "
+                        "or weaker if you wish to allow this transaction to proceed anyway.");
+                case AddressResolutionError::UnifiedAddressResolutionError:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "Could not select a unified address receiver that allows this transaction "
+                        "to proceed without publicly revealing the transaction amount. THIS MAY AFFECT "
+                        "YOUR PRIVACY. Resubmit with the `privacyPolicy` parameter set to "
+                        "`AllowRevealedAmounts` or weaker if you wish to allow this transaction to "
+                        "proceed anyway.");
```

Comment:
The error message here also is a little confusing relative to the name of the error case.

Reply:
Can you tell me a bit what you’d like to see? From my perspective, they seem aligned. The gist is that there’s a UA recipient and we couldn’t find a sending-pool⇔recipient-receiver combo[^1] that doesn’t require (at least) `RevealedAmounts`.

[^1]: As usual, this is a bit overconstrained. There’s a future PR (deep enough in the graph that it hasn’t been opened yet) where we only ever complain about `RevealedAmounts` if there is actually a valueBalance, allowing multi-pool send to a UA without revealing amounts.

Reply:
BTW, a lot of these messages are changing in the follow-up PR that collects all the policy violations rather than just reporting the first. In particular, this splits into two cases – the first being just if there’s no resolver in the UA (which should be impossible, so more like an assert); the second is we found a resolver (and actually chose it), but it then fails policy validation, which should be a clearer case (and depending on the receiver, a different required policy).

Actually … now I’m curious … look at https://github.com/zcash/zcash/pull/6377/files#diff-7bc4ee2e089be48fc4e40c43202e50ff4297c1d5ce4c47607aea01d128967aa0R270-R311 (might have to expand wallet_tx_builder.cpp first). A UA _requires_ a shielded receiver, right? So, I think the only time we ever select a transparent receiver for a payment is when there’s no Sapling receiver and we’re either before NU5[^1] or we don’t have enough non-Sprout funds to cover the tx. Is that right?

[^1]: I forget – can we even send to UAs before NU5? I _think_ so, but not certain.

Reply:
A UA *must* have a shielded receiver, but there's a nit: it is possible that all of the shielded receivers in a UA are shielded receivers that we don't recognize; for example, if our node has not been updated for some period after a new UA receiver type was introduced.

Reply:
In that case, we might only be able to select the transparent receiver.